### PR TITLE
Test generator

### PR DESCRIPTION
**Pull request by davidbruce**
Adds test generator that can be used to help port over exercises from the Problem Repo.

### CODE REVIEW

**Comment by davidbruce** [comment]
@mfoda 

Feel free to pull this branch down and use this to help stub out new exercises.  There will be some conditions in some canonical-data.jsons that it does not account for, but it should be a good starting point for the majority of exercises.

**Comment by mfoda** [comment]
@davidbruce 
Will do! This should speed things up quite a bit :-)

**Comment by mfoda** [comment]
@davidbruce 

I'm running on Windows and seeing a similar issue as [here](https://github.com/HaxeFoundation/haxe/issues/9481):
```powershell
PS D:\source\haxe\exercism\scripts\testgen>  haxe --run TestGen.hx high-scores
C:\HaxeToolkit\haxe\std/sys/Http.hx:495: characters 4-9 : Uncaught exception X509 - Certificate verification failed, e.g. CRL, CA or signature check failed
C:\HaxeToolkit\haxe\std/sys/Http.hx:65: characters 4-14 : Called from here
C:\HaxeToolkit\haxe\std/sys/Http.hx:251: characters 12-25 : Called from here
C:\HaxeToolkit\haxe\std/sys/Http.hx:68: characters 3-30 : Called from here
C:\HaxeToolkit\haxe\std/sys/Http.hx:497: characters 3-19 : Called from here
TestGen.hx:67: characters 29-62 : Called from here
```
Not sure if it's something with my setup or if it's the github SSL issue mentioned in that ticket.

**Comment by davidbruce** [comment]
That is definitely an issue that I can reproduce on windows.  For now I tried the following successfully:

`haxe.exe --neko testgen.n --main  TestGen.hx`

Then to run:
`neko testgen.n anagram`

I imagine any system target should work.

**Comment by davidbruce** [comment]
Also if you do use this and you are using VScode you will more than likely want to run format the Test.hx.  I'll probably setup a global haxe-formatter config this weekend.  I'll then probably have that be run as a part of the generator.

**Comment by mfoda** [comment]
> That is definitely an issue that I can reproduce on windows. For now I tried the following successfully:
> 
> `haxe.exe --neko testgen.n --main TestGen.hx`
> 
> Then to run:
> `neko testgen.n anagram`
> 
> I imagine any system target should work.

Yes it's working now with the above and seems to run great :-)
Will switch to using this from now on for test creation, where possible.

**Comment by BNAndras** [comment]
> Also if you do use this and you are using VScode you will more than likely want to run format the Test.hx.  I'll probably setup a global haxe-formatter config this weekend.  I'll then probably have that be run as a part of the generator.  

@Nezteb, is this haxe-formatter config something easily doable? It'd be great to add that functionality before we review and merge this PR.

File: scripts/testgen/TestGen.hx [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+    static var testCaseTmpl = new Template(
+        "    it(\"::testCase::\", {
+                ::testCriteria::
+            });
+        "
+    );
+
+    public static function main() {
+        var dataUrl            = new Template("https://raw.githubusercontent.com/exercism/problem-specifications/master/exercises/::exercise::/canonical-data.json");
+        var descriptionUrl     = new Template("https://raw.githubusercontent.com/exercism/problem-specifications/master/exercises/::exercise::/description.md");
```
Comment (BNAndras) [inline scripts/testgen/TestGen.hx:52]: `./bin/configlet sync --docs --filepaths --metadata --tests include -uye <slug>` would create the folders and files we need. We'd still need to fetch the canonical data json but we might be able to get it from configlet's cached copy of the problem-specifications repo.

**Comment by Nezteb** [comment]
Okay @BNAndras, the formatter is added. I'll have to figure out why CI is failing though.

The reason for the massive PR now is because commit 39c92e9 formatted pretty much every file.

**Comment by Nezteb** [comment]
Ah crud I didn't mean to close this. I was fixing the commits I had made with no author username/email... #GitIsHard

And now I'm unable to re-open the PR for some reason.

I'll try to open a new PR and salvage some of the commits... 😓

**Comment by BNAndras** [comment]
Fine by me as long as there's a link back to this PR for tracking purposes.

**Pull request closed without merging**